cp_diffdock_api.py

```diff
@@ -0,0 +1,261 @@
+    This contains the data for the DiffDock csv input file.
+    """
+    name: str
+    protein: PreppedComplexUnit
+    ligand: PreppedComplexUnit
+
+    def __iter__(self):
+        # Specific about fields here to avoid converting protein and ligand to dict
+        return iter([self.name, self.protein, self.ligand])
+
+
+@dataclass
+class PreppedRequest:
+    """
+    A DiffDock request after preparing the pdb, sdf, and csv input files.
+    """
+    csv_file: str
+    out_dir: str
+    complex_infos: List[PreppedComplex]
+
+    def __iter__(self):
+        # Specific about fields here to avoid converting complex_infos to dict
+        return iter([self.csv_file, self.out_dir, self.complex_infos])
+
+@dataclass
+class DiffDockOptions:
+    work_dir: Path
+
+    @staticmethod
+    def make(baseDir=None, requestId=None):
+        # get_id needs to run every time, so don't use default arguments
+        if baseDir is None:
+            baseDir = f"/tmp/diffdock{os.getpid()}"
+        if requestId is None:
+            requestId = requestIdGenerator.get_id()
+        work_dir = Path(baseDir) / f"dock_{requestId}"
+        return DiffDockOptions(
+            work_dir=work_dir
+        )
+
+
+class DiffDockNothingToDoException(Exception):
+    pass
+
+class DiffDockSetupException(Exception):
+    pass
+
+
+class DiffDockApi:
+    """
+    Wrapper for running DiffDock with DiffDockProtocol objects for requests and responses.
+    """
+    @staticmethod
+    def run_diffdock(request: DiffDockProtocol.Request, options: DiffDockOptions=None) -> DiffDockProtocol.Response:
+        try:
+            log.info("diffdock preparing request...")
+            preppedRequest = DiffDockApi.prepare_request(request, options)
+
+            args = InferenceArgs({
+                "protein_ligand_csv": preppedRequest.csv_file,
+                "out_dir": preppedRequest.out_dir,
+                "add_hs": request.add_hs,
+                "keep_hs": request.keep_hs,
+                "keep_src_3d": request.keep_src_3d,
+                "samples_per_complex": request.samples_per_complex,
+            })
+            log.info("diffdock running inference...")
+            RunInference(args)
+            response = DiffDockApi.process_results(preppedRequest)
+            log.info("diffdock prepared response.")
+            return response
+        except DiffDockNothingToDoException:
+            return DiffDockProtocol.Response.makeResults()
+        except Exception as ex:
+            return DiffDockProtocol.Response.makeError(f"DiffDock failed. {ex}")
+
+
+    @staticmethod
+    def prepare_request(docking_request: DiffDockProtocol.Request, options: DiffDockOptions=None) -> PreppedRequest:
+        """
+        DiffDock works on files in directories. This function prepares the pdb, sdf, and csv files needed
+        to fulfill the DiffDock request.
+        """
+
+        if len(docking_request.proteins) == 0 or len(docking_request.ligands) == 0:
```

Comment:
I'm not sure I buy it, but the pythonic-style guidelines suggest using truthiness instead of testing len for zero. 
`   if not docking_request.proteins or not docking_request.ligands:`